--- mod_rpaf-2.0.c	2008-01-01 04:05:40.000000000 +0100
+++ mod_rpaf-2.0.c.oden	2011-05-27 08:46:55.295769345 +0200
@@ -72,6 +72,8 @@
 #include "http_vhost.h"
 #include "apr_strings.h"
 
+#include <arpa/inet.h>
+
 module AP_MODULE_DECLARE_DATA rpaf_module;
 
 typedef struct {
@@ -155,6 +157,8 @@ static apr_status_t rpaf_cleanup(void *d
 static int change_remote_ip(request_rec *r) {
     const char *fwdvalue;
     char *val;
+    apr_port_t tmpport;
+    apr_pool_t *tmppool;
     rpaf_server_cfg *cfg = (rpaf_server_cfg *)ap_get_module_config(r->server->module_config,
                                                                    &rpaf_module);
 
@@ -185,6 +189,11 @@ static int change_remote_ip(request_rec
             apr_pool_cleanup_register(r->pool, (void *)rcr, rpaf_cleanup, apr_pool_cleanup_null);
             r->connection->remote_ip = apr_pstrdup(r->connection->pool, ((char **)arr->elts)[((arr->nelts)-1)]);
             r->connection->remote_addr->sa.sin.sin_addr.s_addr = apr_inet_addr(r->connection->remote_ip);
+            tmppool = r->connection->remote_addr->pool;
+            tmpport = r->connection->remote_addr->port;
+            memset(r->connection->remote_addr, '\0', sizeof(apr_sockaddr_t));
+            r->connection->remote_addr = NULL;
+            apr_sockaddr_info_get(&(r->connection->remote_addr), r->connection->remote_ip, APR_UNSPEC, tmpport, 0, tmppool);
             if (cfg->sethostname) {
                 const char *hostvalue;
                 if (hostvalue = apr_table_get(r->headers_in, "X-Forwarded-Host")) {
